hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/StreamFilesPTest.java

```diff
@@ -16,295 +16,211 @@
+    public void when_watchedDirDeleted_then_complete() throws Exception {
+        // Given
+        initializeProcessor();
+        driverThread.start();
 
-        RandomWriter(File file, int numLines, String prefix) {
-            this.file = file;
-            this.numLines = numLines;
-            this.prefix = prefix;
-        }
+        // When
+        assertTrue(workDir.delete());
 
-        @Override
-        public void run() {
-            try (FileOutputStream fos = new FileOutputStream(file);
-                    OutputStreamWriter osw = new OutputStreamWriter(fos, StandardCharsets.UTF_8);
-                    BufferedWriter bw = new BufferedWriter(osw)
-            ) {
-                for (int i = 0; i < numLines; i++) {
-                    bw.write(prefix + " " + i + " Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua\n");
-                    bw.flush();
-                    osw.flush();
-                    fos.flush();
-                    if (random.nextInt(100) < 5)
-                        LockSupport.parkNanos(1_000_000); // 1ms
-                }
-            } catch (IOException e) {
-                throw sneakyThrow(e);
-            }
-        }
+        // Then
+        assertTrueEventually(() -> assertTrue(completedNormally));
     }
 
-    private DAG buildDag() {
-        DAG dag = new DAG();
-        Vertex reader = dag.newVertex("reader", Processors.streamFiles(directory.getPath()))
-                           .localParallelism(1);
-        Vertex writer = dag.newVertex("writer", writeList(list.getName())).localParallelism(1);
-        dag.edge(between(reader, writer));
-        return dag;
-    }
-
-    private File createNewFile() {
-        File file = new File(directory, randomName());
-        assertTrueEventually(() -> assertTrue(file.createNewFile()));
-        return file;
+    private void driveProcessor() {
+        while (!completedNormally && !interrupted()) {
+            completedNormally = processor.complete();
+            updateFileOffsetsSize();
+        }
+        completedNormally = true;
     }
 
-    private static void appendToFile(File file, String... lines) throws IOException {
-        try (PrintWriter writer = new PrintWriter(new FileOutputStream(file, true))) {
-            for (String payload : lines) {
-                writer.write(payload + '\n');
-            }
-        }
+    private void updateFileOffsetsSize() {
+        fileOffsetsSize = processor.fileOffsets.size();
     }
 
-    private static File createTempDirectory() throws IOException {
-        Path directory = Files.createTempDirectory("read-file-stream-p");
-        File file = directory.toFile();
-        file.deleteOnExit();
-        return file;
+    private void assertEmittedCountEventually(long expected) {
+        assertTrueEventually(() -> assertTrue(emittedCount >= expected));
+        assertEquals(expected, emittedCount);
     }
```

Comment:
It will block for a shorter time in the case of a test failure that manifests as too many items.